# Cleaning Up Data

## Data is from Kaggle
https://www.kaggle.com/datasets/becksddf/churn-in-telecoms-dataset/data

# Imports

In [1]:
#Loading all imports for what I'm trying to accomplish
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats


## <span style="color:blue">Load data</span>

In [2]:
churndata = pd.read_csv('../datasets/churn.csv')

In [3]:
churndata.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

# <span style="color:blue">Looking at the head we have some True & False and some Yes & no... we will want to change this later on</span>

In [4]:
churndata.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [5]:
churndata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

### Looking over the uploaded data using`.head()`, `.describe()`, `.info()`.

# <span style="color:blue">Lets go ahead and drop the phone nuumber column as I dont see any benefit from keeping it</span>

In [6]:
churndata.drop(columns='phone number', inplace=True) 

In [7]:
churndata.columns

Index(['state', 'account length', 'area code', 'international plan',
       'voice mail plan', 'number vmail messages', 'total day minutes',
       'total day calls', 'total day charge', 'total eve minutes',
       'total eve calls', 'total eve charge', 'total night minutes',
       'total night calls', 'total night charge', 'total intl minutes',
       'total intl calls', 'total intl charge', 'customer service calls',
       'churn'],
      dtype='object')

<span style="color:blue">We successfully removed phone numbers from columns</span>

<span style="color:blue">I also think its a good idea to take a look at State to make sure we have all 50 and no more than that</span>


In [8]:
len(churndata['state'].value_counts())

51

<span style="color:blue">Hmm.. 51 lets look at this</span>

In [9]:
churndata['state'].value_counts().sort_index()

AK     52
AL     80
AR     55
AZ     64
CA     34
CO     66
CT     74
DC     54
DE     61
FL     63
GA     54
HI     53
IA     44
ID     73
IL     58
IN     71
KS     70
KY     59
LA     51
MA     65
MD     70
ME     62
MI     73
MN     84
MO     63
MS     65
MT     68
NC     68
ND     62
NE     61
NH     56
NJ     68
NM     62
NV     66
NY     83
OH     78
OK     61
OR     78
PA     45
RI     65
SC     60
SD     60
TN     53
TX     72
UT     72
VA     77
VT     73
WA     66
WI     78
WV    106
WY     77
Name: state, dtype: int64

<span style="color:blue">51 because of DC so we can keep it</span>

In [10]:
columns = list(churndata.columns)
[{(x, (len(churndata[x].value_counts()))): [churndata[x].value_counts()]} for x in columns]  
#Able to view the value counts in each column in a list that contains a dictionary

[{('state',
   51): [WV    106
   MN     84
   NY     83
   AL     80
   WI     78
   OR     78
   OH     78
   VA     77
   WY     77
   CT     74
   ID     73
   VT     73
   MI     73
   UT     72
   TX     72
   IN     71
   KS     70
   MD     70
   NJ     68
   MT     68
   NC     68
   CO     66
   NV     66
   WA     66
   MS     65
   RI     65
   MA     65
   AZ     64
   FL     63
   MO     63
   ND     62
   ME     62
   NM     62
   OK     61
   NE     61
   DE     61
   SC     60
   SD     60
   KY     59
   IL     58
   NH     56
   AR     55
   GA     54
   DC     54
   TN     53
   HI     53
   AK     52
   LA     51
   PA     45
   IA     44
   CA     34
   Name: state, dtype: int64]},
 {('account length',
   212): [105    43
   87     42
   93     40
   101    40
   90     39
          ..
   191     1
   199     1
   215     1
   221     1
   2       1
   Name: account length, Length: 212, dtype: int64]},
 {('area code',
   3): [415    1655
   510     840
   408     

In [11]:
churndata['area code'].value_counts(normalize=False), churndata['area code'].value_counts(normalize=True)

(415    1655
 510     840
 408     838
 Name: area code, dtype: int64,
 415    0.496550
 510    0.252025
 408    0.251425
 Name: area code, dtype: float64)

<span style="color:blue">415 represents approximately 49.66% of the total 'area code' values.</span>
    
<span style="color:blue">510 represents approximately 25.20% of the total 'area code' values.</span>
    
<span style="color:blue">408 represents approximately 25.14% of the total 'area code' values.</span>

<span style="color:red">value_counts(normalize=False) how many times each area code shows up</span>

<span style="color:green">value_counts(normalize=True) what fraction of the total area codes each one represents</span>

## To make the modeling easier we need to change <span style="color:green">True & yes</span> and <span style="color:red">False & no</span> 

_1 for <span style="color:green">True & yes</span>_

_0 for <span style="color:red">False & no</span>_

In [18]:
churndata
#international plan Y/N
#voice mail plan Y/N
#churn T/F

state  account length  area code international plan voice mail plan  \
0       KS             128        415                 no             yes   
1       OH             107        415                 no             yes   
2       NJ             137        415                 no              no   
3       OH              84        408                yes              no   
4       OK              75        415                yes              no   
...    ...             ...        ...                ...             ...   
3328    AZ             192        415                 no             yes   
3329    WV              68        415                 no              no   
3330    RI              28        510                 no              no   
3331    CT             184        510                yes              no   
3332    TN              74        415                 no             yes   

      number vmail messages  total day minutes  total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      total day charge  total eve minutes  total eve calls  total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      total night minutes  total night calls  total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      total intl minutes  total intl calls  total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9                 6            

# Churn

In [19]:
churndata.churn.value_counts(), churndata.churn.value_counts(normalize=True) 

(False    2850
 True      483
 Name: churn, dtype: int64,
 False    0.855086
 True     0.144914
 Name: churn, dtype: float64)

## output tells us 
<span style="color:red">False</span> appears 2850 times, making up about 85.51% of the total.

<span style="color:green">True</span> appears 483 times, making up about 14.49% of the total.

In [21]:
churndata.loc[churndata['churn'] == True,'churn'] =1 #reminder 1=true 
churndata.loc[churndata['churn'] == False,'churn'] =0 #reminder 0=false 

churndata['churn'].value_counts()

0    2850
1     483
Name: churn, dtype: int64

# International plan

In [22]:
churndata['international plan'].value_counts(), churndata['international plan'].value_counts(normalize=True)

(no     3010
 yes     323
 Name: international plan, dtype: int64,
 no     0.90309
 yes    0.09691
 Name: international plan, dtype: float64)

## output tells us 
<span style="color:red">no</span> appears 3010 times, making up about 90.31% of the total.

<span style="color:green">yes</span> appears 323 times, making up about 9.69% of the total.


In [25]:
churndata['international plan'] = churndata['international plan'].replace('yes', 1) #reminder 1=yes 
churndata['international plan'] = churndata['international plan'].replace('no', 0) #reminder 0=no 

churndata['international plan'].value_counts()

0    3010
1     323
Name: international plan, dtype: int64

# Voice mail 

In [26]:
churndata['voice mail plan'].value_counts(), churndata['voice mail plan'].value_counts(normalize=True)

(no     2411
 yes     922
 Name: voice mail plan, dtype: int64,
 no     0.723372
 yes    0.276628
 Name: voice mail plan, dtype: float64)

## output tells us 
<span style="color:red">no</span> appears  2411 times, making up about 72.34%  of the total.

<span style="color:green">yes</span> appears 922 times, making up about 27.66% of the total.

In [28]:
churndata['voice mail plan'] = churndata['voice mail plan'].replace('yes', 1) #reminder 1=yes 
churndata['voice mail plan'] = churndata['voice mail plan'].replace('no', 0) #reminder 0=no 

churndata['voice mail plan'].value_counts()

0    2411
1     922
Name: voice mail plan, dtype: int64

In [29]:
churndata

state  account length  area code  international plan  voice mail plan  \
0       KS             128        415                   0                1   
1       OH             107        415                   0                1   
2       NJ             137        415                   0                0   
3       OH              84        408                   1                0   
4       OK              75        415                   1                0   
...    ...             ...        ...                 ...              ...   
3328    AZ             192        415                   0                1   
3329    WV              68        415                   0                0   
3330    RI              28        510                   0                0   
3331    CT             184        510                   1                0   
3332    TN              74        415                   0                1   

      number vmail messages  total day minutes  total day calls  \
0                        25              265.1              110   
1                        26              161.6              123   
2                         0              243.4              114   
3                         0              299.4               71   
4                         0              166.7              113   
...                     ...                ...              ...   
3328                     36              156.2               77   
3329                      0              231.1               57   
3330                      0              180.8              109   
3331                      0              213.8              105   
3332                     25              234.4              113   

      total day charge  total eve minutes  total eve calls  total eve charge  \
0                45.07              197.4               99             16.78   
1                27.47              195.5              103             16.62   
2                41.38              121.2              110             10.30   
3                50.90               61.9               88              5.26   
4                28.34              148.3              122             12.61   
...                ...                ...              ...               ...   
3328             26.55              215.5              126             18.32   
3329             39.29              153.4               55             13.04   
3330             30.74              288.8               58             24.55   
3331             36.35              159.6               84             13.57   
3332             39.85              265.9               82             22.60   

      total night minutes  total night calls  total night charge  \
0                   244.7                 91               11.01   
1                   254.4                103               11.45   
2                   162.6                104                7.32   
3                   196.9                 89                8.86   
4                   186.9                121                8.41   
...                   ...                ...                 ...   
3328                279.1                 83               12.56   
3329                191.3                123                8.61   
3330                191.9                 91                8.64   
3331                139.2                137                6.26   
3332                241.4                 77               10.86   

      total intl minutes  total intl calls  total intl charge  \
0                   10.0                 3               2.70   
1                   13.7                 3               3.70   
2                   12.2                 5               3.29   
3                    6.6                 7               1.78   
4                   10.1                 3               2.73   
...                  ...               ...                ...   
3328                 9.9      

In [40]:
churndata.isna().sum()

state                     0
account length            0
area code                 0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

In [55]:
churndata.to_csv('../datasets/clean_churn.csv')